core/src/main/java/kafka/server/SharePartitionManager.java

```diff
@@ -63,80 +66,129 @@ public class SharePartitionManager {
-
-                        partitionCacheMap.get(sharePartitionKey(groupId, topicIdPartition))
-                            .acquire(memberId, fetchPartitionData).whenComplete((acquiredRecords, throwable) -> {
-                                ShareFetchResponseData.PartitionData partitionData = new ShareFetchResponseData.PartitionData()
-                                    .setPartitionIndex(topicIdPartition.partition());
-
-                                if (throwable != null) {
-                                    partitionData.setErrorCode(Errors.forException(throwable).code());
-                                } else {
-                                    // Maybe check if no records are acquired and we want to retry replica
-                                    // manager fetch. Depends on the share partition manager implementation,
-                                    // if we want parallel requests for the same share partition or not.
-                                    partitionData
-                                        .setPartitionIndex(topicIdPartition.partition())
-                                        .setRecords(fetchPartitionData.records)
-                                        .setErrorCode(fetchPartitionData.error.code())
-                                        .setAcquiredRecords(acquiredRecords)
-                                        .setAcknowledgeErrorCode(Errors.NONE.code());
-                                }
-                                result.put(topicIdPartition, partitionData);
-                            });
-                    });
-                    future.complete(result);
-                    return BoxedUnit.UNIT;
+        Map<ShareFetchPartitionData, CompletableFuture<Map<TopicIdPartition, ShareFetchResponseData.PartitionData>>>
+                fetchResultData = fetchResults();
+        return fetchResultData.get(shareFetchPartitionData);
+    }
+
+    public void addToFetchQueue(ShareFetchPartitionData shareFetchPartitionData) {
+        lock.lock();
+        try {
+            fetchQueue.add(shareFetchPartitionData);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public boolean isFetchQueueEmpty() {
+        lock.lock();
+        try {
+            return fetchQueue.isEmpty();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public ShareFetchPartitionData pollFetchQueue() {
+        lock.lock();
+        try {
+            return fetchQueue.poll();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    public Map<ShareFetchPartitionData, CompletableFuture<Map<TopicIdPartition, ShareFetchResponseData.PartitionData>>> fetchResults() {
+        Map<ShareFetchPartitionData, CompletableFuture<Map<TopicIdPartition, ShareFetchResponseData.PartitionData>>> fetchResults =
+                new HashMap<>();
+        lock.lock();
+        try {
+            while (!isFetchQueueEmpty()) {
```

Comment:
Seeing the semantics it does mean if share share fetch request comes from member twice then one of them can never be completed as the key in map seems not do that justice.

Reply:
yup, you are correct. Using the approach for locking on individual share partitions that you discussed, this should work